extensions/2.0/Khronos/KHR_materials_transmission/README.md

```diff
@@ -0,0 +1,129 @@
+
+## Limitations of Alpha as Coverage
+Alpha-as-coverage can basically be thought of as describing whether the surface exists or not. When alpha is 0, nothing is rendered while alpha=1 means that the material is solid. It's most useful for materials such as gauze or burlap that only partially cover the area in which the material is rendered. These materials contain lots of small holes where the light can pass through so, strictly speaking, the light isn't actually entering the material itself (i.e. the surface isn't there). For other materials, like clear glass, a photon of light might actually enter the surface and be transmitted through. In this case, the light can be reflected, refracted, absorbed or scattered. Simple alpha-as-coverage cannot represent any of these effects and so we need very different logic to render them correctly.
+
+The most important of these features is surely reflection. Even a completely transparent, infinitely thin, glass material is still visible due to its reflection of light. However, a material using purely alpha as coverage is not visible at all when alpha = 0 (because the surface effectively doesn't exist).
+
+<figure>
+<img src="./figures/OpacityComparison.png"/>
+<figcaption><em>White sphere with alpha coverage of 25% (left) vs the same sphere with 100% optical transparency (right). Note that alpha coverage of 25% is shown because 0% would be completely invisible. Also note that alpha-as-coverage reflects light in proportion to its opacity. i.e. the more transparent it is, the weaker the reflections will appear.</em></figcaption>
+</figure>
+
+The `KHR_materials_transmission` extension provides a way to define glTF 2.0 materials that are transparent to light in a physically plausible way. That is, it enables the creation of transparent materials that absorb, reflect and transmit light depending on the incident angle and the wavelength of light. Common uses cases for thin-surface transmissive materials include plastics and glass.
+
+This extension aims to address the simplest and most common use cases for optical transparency: infinitely-thin materials with no refraction, scattering, or dispersion. Dealing exclusively with “thin” materials (i.e. materials where only the surface is considered and not the volume) allows many simplifications to be made when calculating things like refraction and absorption.
+
+## Extending Materials
+
+A transparent material is defined by adding the `KHR_materials_transmission` extension to any glTF material. When present, the extension indicates that a material should be rendered as a transparent surface and be blended as defined in the following spec. Note that, as rendering transparent objects presents many difficult-to-solve issues with primitive ordering, this extension does not dictate rendering algorithms.
```

Comment:
I think that we should make a painters (depth sorting) algorithm normative - otherwise we risk objects 'popping' in and out.

Reply:
Mike is correct that specific rendering techniques are out of scope for normative glTF spec..  This particular suggestion would prohibit many real-time implementations, which is not what we want.

Separately, 3D Commerce may require something along these lines for certification, but beware that you may be excluding more clients than you bargained for here.

Reply:
I think it could pose a performance problem for a particular type of real-time implementations - mostly webGL / script based ones.
The cost would be performance - however the gain would be predictable results.
Depending on implementation (language and platform) you would get varying degrees of performance penalty, for instance a C based engine could do it on the CPU without too much overhead.
A Vulkan based engine could do it on the GPU.

I would not want us to limit this due to current viewer implementation bottlenecks - it could be a good motivator to push development forward.

Reply:
The rendering should not be normative.
Depending on the rendering method (forward vs. deferred plus it's variations), different and better algorithms do exist.
E.g. pixel sorting (aka order independent transparency) should not be excluded.
From my point of view, we should point to non-normative approaches.

Reply:
@rsahlin Also consider that many glTF implementations exist within rendering libraries that are used by 3rd-party applications.  The library cannot dictate how the application structures its render passes or loops.  That's part of what makes this extension particularly challenging, as it imposes some nontrivial best practices there.  These should be as clearly spelled out as we can manage, but not marked normative.  Little more than the glTF file structure and its meaningful interpretation is normative to glTF.

Reply:
I think it is a simplification calling this the 'rendering ' - to me it boils down to what is visible through the primitives using transmission.

If it's the sorting algorithm you object to then I would suggest a wording similar to:
'The background and other objects, including primitives using transmission, should be visible through the material (that is using transmission)'

Reply:
As discussed on today's call, Sketchfab's refraction material has a similar baseline as what is proposed here as an implementation note: we only require the opaque scene to be sampled by the transmission material.  From their [documentation ](https://help.sketchfab.com/hc/en-us/articles/202602073-Transparency):
- A refractive object can not refract itself.
- A refractive object can not refract other refractive objects.
- There can be sorting issues when mixing refractive and transparent objects.
- There can be artifacts when the camera moves.
- Objects outside of the field of view are not refracted.

And for convenience, here are some [sample models](https://sketchfab.com/Sketchfab/collections/refraction-examples) that demonstrate their refraction material.

Sketchfab has established this as an acceptable baseline for their community and I would hope that our spec would be structured in such a way that their implementation is conformant.  It seems like a reasonable baseline.

Reply:
Seems to be a similar situation for Blender's realtime renderer (Eevee). It cannot show refractive objects behind other refractive objects, but can display transparent and opaque objects behind refractive objects. https://blenderartists.org/t/realistic-glass-in-eevee/1149937/29